### ISSUE TITLE
multi-lua install on Windows

### ISSUE THREAD

**Issue by Tieske** [keplerproject/luarocks#403]
I'm trying to setup a multi-Lua (5.1, 5.2, 5.3) installation on Windows and am running into some installer issues/questions.

1. the header files in a standard Lua install are not versioned. I don't know how unix handles this, but I put them in versioned subdirs; e.g. from `.\include\lua.h` to `.\include\lua\5.3\lua.h` (so it matches the `lib` and `share` structure). But the Windows installer doesn't look there. 
2. the default config file written by the windows installer is `config.lua` and not a versioned one. Whereas LuaRocks `cfg` module first looks for a versioned one, and only falls back to an unversioned one if not found.

Question; shouldn't the installer create versioned files/structures by default? and also look in versioned locations first?

**Comment by ignacio** [comment]
I know that's what not you're asking, but, do you intend to use the same LuaRocks install with different config files?

Regarding your question, if a versioned file is created by default, what should we do with the unversioned one? Ignore it? Use it so defaults can be set for all versions?

**Comment by siffiejoe** [comment]
The unversioned one is ignored if a versioned one is available (and the version matches). This is what happens now already. Creating versioned configuration files for new installs just makes it easier to make the LuaRocks installation work with multiple Lua versions later. My LR4Win project uses versioned configuration files only. The rocks subdirectory should be versioned by default too, while we are at it ...

**Comment by ignacio** [comment]
So a change there can break existing installs.
Should the installer copy the existing config into a versioned scheme? A big fat warning should be printed? Or this change should be made with in, dunno, 2.3 or 3.0 ?

**Comment by ignacio** [comment]
@Tieske You way is almost like what Debian does (dunno about other distros):

~~~~
/usr/include/lua5.1
/usr/include/lua5.2
~~~~

The lib folder is not versioned, though.

**Comment by siffiejoe** [comment]
> So a change there can break existing installs.

How so? I haven't used the official Windows installer for some time, but it used to refuse to install into a non-empty target directory. So old installations are totally unaffected by this change, and new installations would only use the versioned configuration files from now on.

**Comment by ignacio** [comment]
Yes, you're right.
I was mostly thinking about Linux users. What happens when you upgrade LuaRocks by itself? Or if you're using the package manager, what happens with config files ? Those were the kind of scenarios that I had in mind.

**Comment by siffiejoe** [comment]
On Linux the `site_config.lua` is installed in a versioned directory, and `config.lua` is versioned by default already.

**Comment by ignacio** [comment]
Well, the debian package install the system config in /etc/luarocks/config.lua unless I'm mistaken.

**Comment by hishamhm** [comment]
> Question; shouldn't the installer create versioned files/structures by default? and also look in versioned locations first?

Yes and yes. The Unix build was changed to do this a while ago.

> What happens when you upgrade LuaRocks by itself? 

Upgrading LuaRocks only upgrades the "source code" .lua files, the configuration is kept untouched (including `site_config.lua`)

> Or if you're using the package manager, what happens with config files ? 

This is up to the package manager to decide, and I can't speak for each of them (except for GoboLinux ;-) ), but they usually keep user-modified config files around.

How and if they transition from `config.lua` to `config-5.x.lua` is up for them to decide too. The Unix makefile that we ship now only writes a `/etc/luarocks/config-5.x.lua` file if it doesn't exist, and does not check at all if `/etc/luarocks/config.lua` exists.

**Comment by ignacio** [comment]
Ok. Fine about package managers.
Regarding the luarocks upgrade scenario, so if site_config is not touched, then everything still works, right?

**Comment by hishamhm** [comment]
> Regarding the luarocks upgrade scenario, so if site_config is not touched, then everything still works, right?

Well, that's the idea :)

**Comment by ignacio** [comment]
Thanks! (and sorry for the possible noise)

**Comment by Tieske** [comment]
Ok, I prepared a fix to install versioned files by default, and look for those first also.

But setting it up properly versioned entails some more; 

1. The installer always versiones LuaRocks itself as well. Set installation for `C:\LuaRocks\` and it will install in `C:\LuaRocks\2.2\`. This is the actual cause of issue #151. Shouldn't we drop this version number in the installation path?
2. config files don't match; after installing for 5.1 and 5.3 (in that order, with my fix) I have the following files;

File: `site_config.lua`
```lua
local site_config = {}
site_config.LUA_INCDIR=[[C:\\Lua\include\lua\5.3]]
site_config.LUA_LIBDIR=[[C:\\Lua\bin\]]
site_config.LUA_BINDIR=[[C:\\Lua\bin\]]
site_config.LUA_INTERPRETER=[[lua53.exe]]
site_config.LUAROCKS_UNAME_S=[[WindowsNT]]
site_config.LUAROCKS_UNAME_M=[[x86]]
site_config.LUAROCKS_SYSCONFIG=[[C:\\LuaRocks\config-5.3.lua]]
site_config.LUAROCKS_ROCKS_TREE=[[C:\\Lua]]
site_config.LUAROCKS_PREFIX=[[C:\\LuaRocks]]
site_config.LUAROCKS_DOWNLOADER=[[wget]]
site_config.LUAROCKS_MD5CHECKER=[[md5sum]]
return site_config
````
File: `config-5.1.lua`
```lua
rocks_trees = {
    { name = [[user]],
         root    = home..[[/luarocks]],
    },
    { name = [[system]],
         root    = [[C:\\Lua]],
    },
}
variables = {
    MSVCRT = 'MSVCRT',
    LUALIB = 'lua51.dll'
}
verbose = false   -- set to 'true' to enable verbose output
````
Issues:
- `LUAROCKS_SYSCONFIG=[[C:\\LuaRocks\config-5.3.lua]]` will prevent the `config-5.1.lua` file from ever being loaded. So I must override 5.3 settings in `site_config.lua`, but can't...
- if we version the config files, then all version specific settings (`LUA_INCDIR`, `LUA_INTERPRETER`, `LUAROCKS_UNAME_M`, etc) must be located (also) in `config.lua` to override the site settings

So what needs to be done imho:

1. stop versioning LuaRocks itself to fix #151 
2. do not write `LUAROCKS_SYSCONFIG` in `site_config` unless the commandline parameter for this setting is passed to the installer (or a self-contained installation is created)
3. in a versioned `config-x.x.lua` override all version specific settings

What do you think?

**Comment by siffiejoe** [comment]
`site_config.lua` can be versioned as well (`site_config_5_1.lua`, etc.). That's what I do in my Windows installer package. On Unix this isn't really necessary, because the LuaRocks modules are installed in versioned directories by default. Removing `LUAROCKS_SYSCONFIG` should work as well though, since `config-x.y.lua` is a default name anyway.

**Comment by Tieske** [comment]
Yes, that would also be an option, but then the installer should no longer remove an existing installation before installing a new one (as that would delete the existing `site_config` files).

having said that, maybe we should split the current installer in an installer part and a configuration part. So you install it once, and then configure it twice (eg. 5.1 and 5.3 for example)

**Comment by Tieske** [comment]
I update #407 to also version the site_config file. But for now, upon an installation, the existing ones (for other versions) get removed.

But number 2 and 3 of my [todo list](https://github.com/keplerproject/luarocks/issues/403#issuecomment-117988369) are fixed by this.

What's left;

1. stop versioning LuaRocks itself to fix #151 
2. retain existing `site_config` files upon an installation

Thoughts?

**Comment by ignacio** [comment]
Just curious what would be the process to setup LuaRocks with multiple Lua versions then. Your "install / config" seems interesting. You "install" LuaRocks to a given location, and each "config" action can add support for 5.1, 5.2, etc. ?

What about bin scripts?  Wouldn't they step on each other?

**Comment by Tieske** [comment]
The difference per Lua version in the installation is the `site_config_x_x.lua` module and the `config-x.x.lua` file. And additionally the batch files maybe for some hardcodes stuff in there, so probably use a `luarocks51.bat` and a `luarocks53.bat`.

So the installer would install LuaRocks (maybe based on a single Lua version, so including one config?) and the config command would add another set of those 2 config files and a batch file.

The bin scripts is another issue, but I doubt whether it should be considerd a LR issue. If I install a commandline script (say `pe-parser`, for both 5.1 and 5.3, then obviously the last one installed would be used. If I use different bin directories, then the precedence in PATH sets which one is used. There is no way both can be used at the same time.
Another way is to use a single bin folder for the Luas (`\bin\lua51.exe`, `\bin\lua53.exe`, etc in a single bin directory), and use versioned ones for the LR bin scripts (`bin\5.1\busted.bat` and `bin\5.3\busted.bat`). This way they don't step on each others toes, and the user can set precedence using the system path.

**Comment by Tieske** [comment]
> Another way is to use a single bin folder for the Luas ( \bin\lua51.exe ,  \bin\lua53.exe , etc in a single bin directory)

just so support dll's, like OpenSSL etc, need to be copied only once, for all installed Lua versions. And all versions are commandline accessible through their version suffix, without changing the system path.

**Comment by ignacio** [comment]
On Thu, Jul 2, 2015 at 11:46 AM, Thijs Schreijer <notifications@github.com>
wrote:

> The difference per Lua version in the installation is the
> site_config_x_x.lua module and the config-x.x.lua file. And additionally
> the batch files maybe for some hardcodes stuff in there, so probably use a
> luarocks51.bat and a luarocks53.bat.
>
Good. I do have three different LR installs (one for each Lua version) and
drop versioned batch files to use each of them. This change fits my
workflow.

 So the installer would install LuaRocks (maybe based on a single Lua
version, so including one config?) and the config command would add another
set of those 2 config files and a batch file.

:+1:

> The bin scripts is another issue, but I doubt whether it should be
> considerd a LR issue. If I install a commandline script (say pe-parser,
> for both 5.1 and 5.3, then obviously the last one installed would be used.
> If I use different bin directories, then the precedence in PATH sets which
> one is used. There is no way both can be used at the same time.
> Another way is to use a single bin folder for the Luas (\bin\lua51.exe,
> \bin\lua53.exe, etc in a single bin directory), and use versioned ones
> for the LR bin scripts (bin\5.1\busted.bat and bin\5.3\busted.bat). This
> way they don't step on each others toes, and the user can set precedence
> using the system path.
>
Another way could be to always drop an additional versioned batch file
(pe-parser51, etc). It seems less cumbersome that the need to fiddle with
paths.

**Comment by hishamhm** [comment]
> The bin scripts is another issue

We largely ignore this issue on Unix. If the user cares about having separate bin dirs, they can configure it as such. In the default case, we just want scripts to be PATH-accessible. In the general case, if you're running a command-line utility, does it matter if it's running on lua 5.1 or 5.3? When it does make a difference, the rock can provide a versioned script (like luarocks itself does with `luarocks-5.x`). But I don't think we should generate versioned scripts every time.

**Comment by ignacio** [comment]
It makes sense. They should be functionally equivalent so no harm in "last install wins".

**Comment by siffiejoe** [comment]
> retain existing site_config files upon an installation

Actually, I wouldn't do that. If you find an unversioned configuration file in the new installer, just wipe the whole directory like before. Only if the installation supports multiple Lua versions, you can install/replace the version specific configuration files.

> luarocks51.bat and a luarocks53.bat.

The designated name is `luarocks-x.y.bat`. I changed the batch scripts in my installer specifically to match that pattern.

**Comment by daurnimator** [comment]
On 3 Jul 2015 4:42 am, "Hisham Muhammad" <notifications@github.com> wrote:
>>
>> The bin scripts is another issue
>
> We largely ignore this issue on Unix. If the user cares about having
separate bin dirs, they can configure it as such. In the default case, we
just want scripts to be PATH-accessible. In the general case, if you're
running a command-line utility, does it matter if it's running on lua 5.1
or 5.3? When it does make a difference, the rock can provide a versioned
script (like luarocks itself does with luarocks-5.x). But I don't think we
should generate versioned scripts every time.
>
> —
> Reply to this email directly or view it on GitHub.

I wish we did care.  E.g. For busted, every time I want to test across all
lua versions, I have to install busted again for that version

**Comment by siffiejoe** [comment]
> I wish we did care. E.g. For busted, every time I want to test across all lua versions, I have to install busted again for that version

Separate bin directories wouldn't work well either, because you probably would add all of them to your PATH, so you need some env fiddling to select a specific version. Adding a version suffix to every script isn't good either, because only certain types of scripts need to run under different Lua versions (e.g. unit testing, repls, debuggers, ...). One possible way to address this issue would be to use a suffix of "-x.y" instead of "~" when a bin script is replaced.

In the meantime this issue can be worked around by the rockspec author (by providing versioned scripts and one rockspec per Lua version).

**Comment by Tieske** [comment]
I think that versioned bin directories are very useful for advanced scenarios, like multi-version testing etc.

Versioning the individual scripts is something I wouldn't like, seems messy, needs updating the manifest (?), and any calling script (Travis or AppVeyor for example) now needs to be aware of which version to use, which defeats the whole concept of versioning in the first place.

Not having them doesn't give the user an option. Having them for an occasional user does not make a difference for that user, he has to add a single path, just as with unversioned bin directories to his system path. For advanced uses, it provides the possibility of separation, users who don't want it can still add all versiosn to their systempath, whereas those who do need it can use it as such.
One can use a single calling script, with just a different system path set, to make it use another version, without modifications to that calling script.

I'd like versioned bin directories...

and unless I'm missing something, this is an installer feature, not a LR feature. So it could be implemented with a simple commandline switch.

**Comment by ignacio** [comment]
For the record, what I was proposing was having a versioned script in addition to an unversioned one (so you'd get busted.bat and busted-5.1.bat).
But I didn't considered that the manifest has to be updated, so that might be a stopper for this.

Calling scripts *do* some times need to be aware of the version to use (as mentioned by siffiejoe, unit testing, repls, etc).

But, anyway, versioned bin directories seems better.

**Comment by siffiejoe** [comment]
> But, anyway, versioned bin directories seems better.

In my opinion versioned bin directories are pretty much useless. If you don't put them into your PATH, `~/.self/programs/luarocks/bin-5.1/busted x y z` is actually worse than e.g. `lua5.1 -lbusted` (which is already versioned by default on my OS), and you have to do this for *all* installed scripts, versioned or not, and you have to remember which LuaRocks installation has that module installed). And if you do put them into your PATH, then you have to start fiddling with your environment every time you want to use a different Lua version (and when I unit test, I test all three Lua versions in quick succession). Although I guess you could use `luarocks path` for that (which I currently don't use at all) -- or you just reinstall busted for the Lua version you want to test ...

But as I mentioned, you can work around that issue as a rockspec author, so there's no reason for panic.

**Comment by Tieske** [comment]
Reinstalling the modules you need in between test runs is not a clean solution imo. Basically it forces you to change the host system to perform a test. If you use a test script that exits on a first failure for example, you leave the system in an undetermined state. Unless you take precautions of course, but anyway it feels like a 'sloppy' way of working. Setting a path before a run, and not messing with installed components is a lot cleaner imo.

**Comment by Tieske** [comment]
@hishamhm maybe best to give us your verdict on changing this in the defaults or not. So we can leave the bikeshed alone :smile:

**Comment by ignacio** [comment]
No... why? We've almost settled on the colour.

**Comment by siffiejoe** [comment]
Ok, so let's make sure we're even talking about the same building:

This feature is not intended for automatic CI unit testing (à la Travis), because in this case you basically get a fresh module install, a fresh LuaRocks install, and even a fresh Linux install for each Lua version. There will never be `busted-5.1` *and* `busted-5.2` in the same test run (at least that's the most common setup now, IIRC).

So this feature is more for interactive use like:

```sh
$ vim mymodule.lua
$ testy-5.1 mymodule.lua
120 tests ok
$ testy-5.2 mymodule.lua
120 tests ok
$ testy-5.3 mymodule.lua
120 tests ok
$ # ok, I didn't mess anything up ...
$ git add mymodule.lua
$ git commit
```

And, IMHO, this is superior to

```sh
$ vim mymodule.lua
$ /usr/local/lua5.1/bin/testy mymodule.lua
120 tests ok
$ /usr/local/lua5.2/bin/testy mymodule.lua
120 tests ok
$ /usr/local/lua5.3/bin/testy mymodule.lua
120 tests ok
$ # ok, I didn't mess anything up ...
```

I'm not even sure we can just invent directory names like that on Unix.

Relying on modifying PATH would be even worse, because you wouldn't catch the case where the test script isn't installed for all Lua versions. Of course you could just add helper scripts (`testy-5.x.sh`) to your projects (I'd probably use aliases) ...

I'm not sure we can expect a typical LuaRocks user to figure out which rocks tree a specific module was installed to, locate the bin directory that belongs to that tree (configured in the `config-5.x.lua` config file), and call the script via an absolute path. At least now

```sh
$ luarocks-5.2 install busted
$ busted mymodule.lua
```

will definitely run the 5.2 version [*]. With separate bin directories it might run any installed version depending on the order of the PATH.

That said, I can already do anything I need regarding versioned unit testing. I just wanted to point out that the "separate bin dirs"-workaround is not an improvement compared to the "separate rockspecs"-workaround we already have.

> I think that versioned bin directories are very useful for advanced scenarios, like multi-version testing etc.

I don't consider multi-version testing an advanced scenario. Most rockspecs (2019 of 2241) in the repository claim to run on more than one Lua version, so obiously all those modules are tested with Lua 5.1, 5.2, and/or 5.3 on a regular basis ;-p

  [*]: Actually, if you use multiple rocks trees, that could cause problems already. But I have disabled the user rocks tree in all my installations.

**Comment by ignacio** [comment]
@siffiejoe And in your testy (or busted) example, who's responsible for providing those versioned scripts? You (the module author) or LuaRocks?

**Comment by siffiejoe** [comment]
The rockspec author. Only the rockspec (or module) author can decide whether it makes sense to have multiple different versions of a script. We had this topic a short while ago on the [LuaRocks mailing list][1]: You provide one rockspec for each Lua version (using "lua ~> 5.x" as a dependency) and install a versioned script in addition to/instead of the normal non-versioned script. See [here][2]. This way is a bit awkward (and it's definitely a workaround), but as long as the number of supported Lua versions is small, it works just fine.

  [1]: http://comments.gmane.org/gmane.comp.lang.lua.luarocks/5719
  [2]: https://github.com/siffiejoe/lua-testy/tree/master/rockspecs

**Comment by ignacio** [comment]
Thanks. Well, we agree that it is awkward requiring authors to deal with this. imho, it is a bit too much to ask.

Meanwhile, there are some changes in #407 that are valuable and could land. Maybe if [this change](https://github.com/keplerproject/luarocks/pull/407/files#diff-e2f5d08db7fdbe4af160e8aeb68e6ee9R363) is dropped in the meantime it can be merged?

**Comment by Tieske** [comment]
Which part would you like to drop? couldn't tell it from the link.

**Comment by ignacio** [comment]
This part: 
`for _, e in ipairs{ S([[\include\lua\$LUA_VERSION]]), S([[\include\$LUA_VERSION]]), [[\]], [[\include\]]} do`

leave it at;
`for _, e in ipairs{ [[\]], [[\include\]]} do`

**Comment by Tieske** [comment]
if versioned files are to be preferred over non-versioned files (already the case on unix iirc), then I would consider that change a bugfix and leave it in.

A new PR for dropping the 2.2 from the LR install dir, and at the same time retain the config files (both config-x.x.lua and the site_config_x_x.lua) makes sense.

**Comment by siffiejoe** [comment]
IMHO, the location of Lua include files is out of scope for LuaRocks. If you think `include\lua\5.1` is a common subdirectory for Lua header files on Windows, then by all means add it to the list. The sub-directories searched in the Unix configure script are based on the common Lua header file locations in popular Linux distros (most of which do have versioned sub-directories). On Windows there is no common practice, AFAIK, but "everything in one directory" (the `/` case) and plain `include` sub-directories would be my first guesses as well.

What I don't get is the change to `luarocks/cfg.lua`. It basically seems to do nothing ...

**Comment by Tieske** [comment]
>  the location of Lua include files is out of scope for LuaRocks

It can't be out of scope, they are needed to compile modules

> On Windows there is no common practice, AFAIK, but "everything in one directory" (the  /  case) and plain  include  sub-directories would be my first guesses as well.

True, there is none. But, if versioned files have precedence over unversioned files, then the order is important, to test versioned first. And these are my best guesses, if there are others I missed, then they need to be added. 

The installer needs some intelligent guesses to get things working quickly, preferably automatically. So combine that with the precedence, this is the only viable solution I can come up with. I tested some multi-lua version installation with the LR installer, and it consistently picks the wrong header files (or doesn't find them) without this change.

> What I don't get is the change to `luarocks/cfg.lua`. It basically seems to do nothing ...

That code doesn't change anything locally, but the value for the default is now different. And that changes the behaviour [here](https://github.com/Tieske/luarocks/blob/pref_versioned/src/luarocks/cfg.lua#L648-659). Output of the plain LR command now shows this;
````

CONFIGURATION
        Lua version: 5.1
        Configuration files:
                System: C:/Program Files (x86)/LuaRocks/config-5.1.lua (ok)
                User  : C:/Users/Thijs/AppData/Roaming/luarocks/config-5.1.lua (not found)

        Rocks trees in use:
                C:/Users/Thijs/AppData/Roaming/luarocks ("user")
                C:/Program Files (x86)/LuaRocks/systree ("system")

C:\Users\Thijs>
````
Notice the configuration file 'not found': it used to be `config.lua`, but is now the versioned alternative.

**Comment by siffiejoe** [comment]
I have used `\include\lua51` in my LuaRocks installer for Windows. It is derived from the Debian way of doing things with the usual Windows-twists applied (the executable is `lua5.1` => `lua51.exe`, the library is `lua5.1.a` => `lua51.dll`, and the include dir is `/usr/include/lua5.1` => `\include\lua51`), although FreeBSD also uses `/usr/local/include/lua51`. You can also try the real Debian path with the dot: `\include\lua5.1`. But I'm not sure how many multi-Lua installations (multiple Lua versions sharing a common directory structure) on Windows are actually out there. A multi-Lua installation is not a prerequisite for a versioned LuaRocks installation, and on WIndows nothing encourages you to create shared directory structures ...

**Comment by Tieske** [comment]
just added the `include\lua51\` and `include\lua5.1\`, as well as incorporating @ignacio remark about the backup filename.

Is this good to go now?

**Comment by ignacio** [comment]
Fine by me.

**Comment by Tieske** [comment]
There is one more thing to do, as @siffiejoe [already mentioned](https://github.com/keplerproject/luarocks/issues/403#issuecomment-116780774);

> The rocks subdirectory should be versioned by default too, while we are at it ...

Here's an example showing the location of the manifest file; `C:\Lua\lib\luarocks\rocks\manifest`. I think this should be updated to `C:\Lua\lib\luarocks\rocks-5.1\manifest` so it is also properly versioned.

I think we can implement it by trying a versioned directory first, and fall back to a non-versioned one. That would be backward compatible.

**Comment by hishamhm** [comment]
> I think we can implement it by trying a versioned directory first, and fall back to a non-versioned one. That would be backward compatible.

This is what already happens in Unix:

https://github.com/keplerproject/luarocks/blob/master/configure#L395-L412

**Comment by Tieske** [comment]
So in the Windows installer (which sets up 2 rocktrees by default), we should;

1. check the configured rocktrees for existence `/lib/luarocks/rocks/`
2. if it exists, use that, if not default to a versioned one
3. write the path to the `LUAROCKS_ROCKS_SUBDIR` variable in the `site_config_X_X.lua` file

in LuaRocks 3;

1. the `cfg` module should default to a versioned path (currently it defaults to an unversioned path)

I'll create a PR for the installer change

**Comment by Tieske** [comment]
closing this as all the elements have been resolved
